File: src/utils/git.ts
import { execSync } from 'node:child_process';

export function isGitRepo(cwd: string): boolean {
	try {
		execSync('git rev-parse --is-inside-work-tree', { cwd, stdio: 'pipe' });
		return true;
	} catch {
		return false;
	}
}

export function getMainBranch(cwd: string): string {
	try {
		const result = execSync(
			'git symbolic-ref refs/remotes/origin/HEAD 2>/dev/null || echo refs/heads/main',
			{ cwd, stdio: 'pipe' },
		);
		const ref = result.toString().trim();
		return ref.replace('refs/remotes/origin/', '').replace('refs/heads/', '');
	} catch {
		return 'main';
	}
}

export function getCurrentBranch(cwd: string): string {
	try {
		return execSync('git branch --show-current', { cwd, stdio: 'pipe' })
			.toString()
			.trim();
	} catch {
		return '';
	}
}

File: internal/util/git.go
package util

import (
	"os/exec"
	"strings"
)

// IsGitRepo checks if cwd is inside a git repository.
func IsGitRepo(cwd string) bool {
	cmd := exec.Command("git", "rev-parse", "--is-inside-work-tree")
	cmd.Dir = cwd
	return cmd.Run() == nil
}

// GetMainBranch returns the main branch name (main or master).
func GetMainBranch(cwd string) string {
	cmd := exec.Command("bash", "-c", "git symbolic-ref refs/remotes/origin/HEAD 2>/dev/null || echo refs/heads/main")
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return "main"
	}
	ref := strings.TrimSpace(string(out))
	ref = strings.TrimPrefix(ref, "refs/remotes/origin/")
	ref = strings.TrimPrefix(ref, "refs/heads/")
	return ref
}

// GetCurrentBranch returns the current git branch name.
func GetCurrentBranch(cwd string) string {
	cmd := exec.Command("git", "branch", "--show-current")
	cmd.Dir = cwd
	out, err := cmd.Output()
	if err != nil {
		return ""
	}
	return strings.TrimSpace(string(out))
}

// GitExec runs a git command in the given directory.
func GitExec(cwd string, args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	cmd.Dir = cwd
	out, err := cmd.CombinedOutput()
	return strings.TrimSpace(string(out)), err
}
